reader.go

```diff
@@ -29,31 +66,847 @@ type Reader struct {
+		})
+		return nil, err
+	}
+
+	r.withLogger(func(l *log.Logger) {
+		l.Printf("sync group finished for group, %v\n", r.config.GroupID)
+	})
+
+	return assignments.Topics, nil
+}
+
+func (r *Reader) rebalance() (map[string][]int32, error) {
+	r.withLogger(func(l *log.Logger) {
+		l.Printf("rebalancing consumer group, %v", r.config.GroupID)
+	})
+
+	if err := r.refreshCoordinator(); err != nil {
+		return nil, err
+	}
+
+	members, err := r.joinGroup()
+	if err != nil {
+		return nil, err
+	}
+
+	subs, err := r.syncGroup(members)
+	if err != nil {
+		return nil, err
+	}
+
+	return subs, nil
+}
+
+func (r *Reader) unsubscribe() error {
+	r.cancel()
+	r.join.Wait()
+	return nil
+}
+
+func (r *Reader) fetchOffsets(subs map[string][]int32) (map[int]int64, error) {
+	conn, err := r.connect(r.stctx)
+	if err != nil {
+		return nil, err
+	}
+	defer conn.Close()
+
+	partitions := subs[r.config.Topic]
+	offsets, err := conn.offsetFetch(offsetFetchRequestV3{
+		GroupID: r.config.GroupID,
+		Topics: []offsetFetchRequestV3Topic{
+			{
+				Topic:      r.config.Topic,
+				Partitions: partitions,
+			},
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	offsetsByPartition := map[int]int64{}
+	for _, pr := range offsets.Responses[0].PartitionResponses {
+		for _, partition := range partitions {
+			if partition == pr.Partition {
+				offset := pr.Offset
+				if offset >= 0 {
+					offset++ // advance to next offset
+				}
+				offsetsByPartition[int(partition)] = offset
+			}
+		}
+	}
+
+	return offsetsByPartition, nil
+}
+
+func (r *Reader) subscribe(subs map[string][]int32) error {
+	// always clear prior subscriptions
+	r.unsubscribe()
+
+	if len(subs[r.config.Topic]) == 0 {
+		return nil
+	}
+
+	offsetsByPartition, err := r.fetchOffsets(subs)
+	if err != nil {
+		return err
+	}
+
+	r.mutex.Lock()
+	r.start(offsetsByPartition)
+	r.mutex.Unlock()
+
+	r.withErrorLogger(func(l *log.Logger) {
+		l.Printf("subscribed to partitions: %+v", offsetsByPartition)
+	})
+
+	return nil
+}
+
+// broker returns a connection to ANY broker
+func (r *Reader) broker() (conn *Conn, err error) {
+	for _, broker := range r.config.Brokers {
+		if conn, err = r.config.Dialer.Dial("tcp", broker); err == nil {
+			return
+		}
+	}
+	return // err will be non-nil
+}
+
+// connect returns a connection to the coordinator for this group
+func (r *Reader) connect(ctx context.Context) (*Conn, error) {
+	r.mutex.Lock()
+	address := r.address
+	r.mutex.Unlock()
+
+	conn, err := r.config.Dialer.DialContext(ctx, "tcp", address)
+	if err != nil {
+		r.withErrorLogger(func(l *log.Logger) {
+			l.Printf("unable to connect to broker, %v\n", address)
+		})
+		return nil, err
+	}
+
+	return conn, nil
+}
+
+func (r *Reader) waitThrottleTime(throttleTimeMS int32) {
+	if throttleTimeMS == 0 {
+		return
+	}
+
+	select {
+	case <-r.stctx.Done():
+		return
+
+	case <-time.After(time.Duration(throttleTimeMS) * time.Millisecond):
```

Comment:
Can we replace the use of `time.After` for a `timer.Timer`? That way we can stop it on return and we don't leak resources if `stctx` is canceled.

Reply:
Sure, happy to change it. 

I figured it wouldn't be a problem since the loop is slow and resources are freed once time.After fires.  It's not so much a leak as holding resources longer than necessary.